### ISSUE TITLE
Does beam-postgres lockingAllTablesFor_ need withRewrittenThread?

### ISSUE THREAD

**Issue by ocharles** [tathougies/beam#279]
I'm trying to write the following:

```haskell
    nextPendingJob
      :: Beam.Q Beam.PgSelectSyntax Database s ( Beam.QExpr Beam.PgExpressionSyntax s JobId.JobId )
    nextPendingJob =
      Beam.all_ ( picofactoryJobs database )
        & Beam.filter_ ( \j -> Job.status j Beam.==. Beam.val_ JobStatus.Pending )
        & Beam.orderBy_ ( Beam.asc_ . Job.scheduledAt )
        & fmap Job.identifier
        & Beam.limit_ 1
        & Beam.lockingAllTablesFor_ Beam.PgSelectLockingStrengthUpdate Nothing
```

But this produces:

```
    • Occurs check: cannot construct the infinite type:
        s ~ Beam.QNested s
      Expected type: Beam.Q
                       Beam.PgSelectSyntax
                       Database
                       (Beam.QNested (Beam.QNested s))
                       (Beam.QGenExpr
                          Beam.QValueContext
                          Beam.PgExpressionSyntax
                          (Beam.QNested (Beam.QNested s))
                          JobId.JobId)
                     -> Beam.Q
                          Beam.PgSelectSyntax
                          Database
                          (Beam.QNested s)
                          (Beam.QExpr Beam.PgExpressionSyntax s JobId.JobId)
        Actual type: Beam.Q
                       Beam.PgSelectSyntax
                       Database
                       (Beam.QNested (Beam.QNested s))
                       (Beam.QGenExpr
                          Beam.QValueContext
                          Beam.PgExpressionSyntax
                          (Beam.QNested (Beam.QNested s))
                          JobId.JobId)
                     -> Beam.Q
                          Beam.PgSelectSyntax
                          Database
                          (Beam.QNested s)
                          (Beam.WithRewrittenThread
                             (Beam.QNested (Beam.QNested s))
                             (Beam.QNested s)
                             (Beam.QGenExpr
                                Beam.QValueContext
                                Beam.PgExpressionSyntax
                                (Beam.QNested (Beam.QNested s))
                                JobId.JobId))
    • In the second argument of ‘(&)’, namely ‘Beam.limit_ 1’
      In the first argument of ‘(&)’, namely
        ‘Beam.all_ (picofactoryJobs database)
           & Beam.filter_
               (\ j -> Job.status j Beam.==. Beam.val_ JobStatus.Pending)
           & Beam.orderBy_ (Beam.asc_ . Job.scheduledAt)
           & fmap Job.identifier
           & Beam.limit_ 1’
      In the expression:
        Beam.all_ (picofactoryJobs database)
          & Beam.filter_
              (\ j -> Job.status j Beam.==. Beam.val_ JobStatus.Pending)
          & Beam.orderBy_ (Beam.asc_ . Job.scheduledAt)
          & fmap Job.identifier
          & Beam.limit_ 1
          & Beam.lockingAllTablesFor_
              Beam.PgSelectLockingStrengthUpdate Nothing
    • Relevant bindings include
        nextPendingJob :: Beam.Q
                            Beam.PgSelectSyntax
                            Database
                            s
                            (Beam.QExpr Beam.PgExpressionSyntax s JobId.JobId)
          (bound at lib/Picofactory/Coordinator/JobQueue/Database.hs:91:5)
   |
96 |         & Beam.limit_ 1
```

However, if I use this `lockingAllTablesFor_` wrapper, everything type checks:

```haskell
lockingAllTablesFor_
  :: forall s db a.
     ( Beam.Database Beam.Postgres db
     , Beam.Projectible Beam.PgExpressionSyntax a
     , Beam.ThreadRewritable ( Beam.QNested s ) a
     )
  => Beam.PgSelectLockingStrength
  -> Maybe Beam.PgSelectLockingOptions
  -> Beam.Q Beam.PgSelectSyntax db (Beam.QNested s) a
  -> Beam.Q Beam.PgSelectSyntax db s ( Beam.WithRewrittenThread ( Beam.QNested s ) s a )
lockingAllTablesFor_ a b c =
  fmap
    ( \a -> Beam.rewriteThread (Proxy @s) a :: Beam.WithRewrittenThread ( Beam.QNested s ) s a )
    ( Beam.lockingAllTablesFor_ a b c )
```

**Comment by tathougies** [comment]
Yes, it does, and it's fixed in 1cdb3fd1c1ae4d467d0417ee5aea4507b4b09447

**Comment by ocharles** [comment]
Cheers!
